src/platform/cmx/chal/shared/cos_types.h

```diff
@@ -0,0 +1,481 @@
+{ return (((cycles_t)c) << TCAP_TIME_QUANTUM_ORD) | TCAP_TIME_MAX_BITS(curr); }
+static inline tcap_time_t
+tcap_cyc2time(cycles_t c) {
+	tcap_time_t t = (tcap_time_t)(c >> TCAP_TIME_QUANTUM_ORD);
+	return t == TCAP_TIME_NIL ? 1 : t;
+}
+#define CYCLES_DIFF_THRESH (1<<14)
+static inline int
+cycles_same(cycles_t a, cycles_t b, cycles_t diff_thresh)
+{
+	return (b < a ? a - b : b - a) <= diff_thresh;
+}
+/* FIXME: if wraparound happens, we need additional logic to compensate here */
+static inline int tcap_time_lessthan(tcap_time_t a, tcap_time_t b) { return a < b; }
+
+typedef enum {
+	TCAP_DELEG_TRANSFER = 1,
+	TCAP_DELEG_YIELD    = 1<<1,
+} tcap_deleg_flags_t;
+
+typedef enum {
+	RCV_NON_BLOCKING = 1,
+	RCV_ALL_PENDING  = 1 << 1,
+} rcv_flags_t;
+
+#define BOOT_LIVENESS_ID_BASE 2
+
+typedef enum {
+	CAPTBL_OP_CPY,
+	CAPTBL_OP_CONS,
+	CAPTBL_OP_DECONS,
+	CAPTBL_OP_THDACTIVATE,
+	CAPTBL_OP_THDDEACTIVATE,
+	CAPTBL_OP_THDTLSSET,
+	CAPTBL_OP_COMPACTIVATE,
+	CAPTBL_OP_COMPDEACTIVATE,
+	CAPTBL_OP_SINVACTIVATE,
+	CAPTBL_OP_SINVDEACTIVATE,
+	CAPTBL_OP_SRETACTIVATE,
+	CAPTBL_OP_SRETDEACTIVATE,
+	CAPTBL_OP_ASNDACTIVATE,
+	CAPTBL_OP_ASNDDEACTIVATE,
+	CAPTBL_OP_ARCVACTIVATE,
+	CAPTBL_OP_ARCVDEACTIVATE,
+	CAPTBL_OP_MEMACTIVATE,
+	CAPTBL_OP_MEMDEACTIVATE,
+	/* CAPTBL_OP_MAPPING_MOD, */
+
+	CAPTBL_OP_MEM_RETYPE2USER,
+	CAPTBL_OP_MEM_RETYPE2KERN,
+	CAPTBL_OP_MEM_RETYPE2FRAME,
+
+	CAPTBL_OP_PGTBLACTIVATE,
+	CAPTBL_OP_PGTBLDEACTIVATE,
+	CAPTBL_OP_CAPTBLACTIVATE,
+	CAPTBL_OP_CAPTBLDEACTIVATE,
+	CAPTBL_OP_CAPKMEM_FREEZE,
+	CAPTBL_OP_CAPTBLDEACTIVATE_ROOT,
+	CAPTBL_OP_PGTBLDEACTIVATE_ROOT,
+	CAPTBL_OP_THDDEACTIVATE_ROOT,
+	CAPTBL_OP_MEMMOVE,
+	CAPTBL_OP_INTROSPECT,
+	CAPTBL_OP_TCAP_ACTIVATE,
+	CAPTBL_OP_TCAP_TRANSFER,
+	CAPTBL_OP_TCAP_DELEGATE,
+	CAPTBL_OP_TCAP_MERGE,
+	CAPTBL_OP_TCAP_WAKEUP,
+
+	CAPTBL_OP_HW_ACTIVATE,
+	CAPTBL_OP_HW_DEACTIVATE,
+	CAPTBL_OP_HW_ATTACH,
+	CAPTBL_OP_HW_DETACH,
+	CAPTBL_OP_HW_MAP,
+	CAPTBL_OP_HW_CYC_USEC,
+	CAPTBL_OP_HW_CYC_THRESH,
+} syscall_op_t;
+
+typedef enum {
+	CAP_FREE = 0,
+	CAP_SINV,		/* synchronous communication -- invoke */
+	CAP_SRET,		/* synchronous communication -- return */
+	CAP_ASND,		/* async communication; sender */
+	CAP_ARCV,               /* async communication; receiver */
+	CAP_THD,                /* thread */
+	CAP_COMP,               /* component */
+	CAP_CAPTBL,             /* capability table */
+	CAP_PGTBL,              /* page-table */
+	CAP_FRAME, 		/* untyped frame within a page-table */
+	CAP_VM, 		/* mapped virtual memory within a page-table */
+	CAP_QUIESCENCE,         /* when deactivating, set to track quiescence state */
+	CAP_TCAP, 		/* tcap captable entry */
+	CAP_HW,			/* hardware (interrupt) */
+} cap_t;
+
+/* TODO: pervasive use of these macros */
+/* v \in struct cap_* *, type \in cap_t */
+#define CAP_TYPECHK(v, t) ((v) && (v)->h.type == (t))
+#define CAP_TYPECHK_CORE(v, type) (CAP_TYPECHK((v), (type)) && (v)->cpuid == get_cpuid())
+
+typedef enum {
```

Comment:
Wow! You seem to be working off of an old commit base. 
This file is cleaned up in the mainline to move hardware specifics to `chal/chal_config.h`.

Reply:
Because many of these files are moved to `cmx` folder, its a pain to manually merge from mainline.

Also, do you need this file to be in `cmx`?

Reply:
the more I think of it, the more I'm confused.

This file is originally in `kernel/include/shared/cos_types.h`. Confused why it's in `cmx` in your case.